src/core/annotation.js

```diff
@@ -227,7 +227,31 @@ function getQuadPoints(dict, rect) {
       quadPointsLists[i].push({ x, y });
     }
   }
-  return quadPointsLists;
+
+  // The PDF specification states in section 12.5.6.10 (figure 64) that the
+  // order of the quadpoints should be bottom left, bottom right, top right
+  // and top left. However, in practice PDF files use a different order,
+  // namely bottom left, bottom right, top left and top right (this is also
+  // mentioned on https://github.com/highkite/pdfAnnotate#QuadPoints), so
+  // this is the actual order we should work with. However, the situation is
+  // even worse since Adobe's own applications and other applications violate
+  // the specification and create annotations with other orders, namely top
+  // left, top right, bottom left and bottom right or even top left, top right,
+  // bottom right and bottom left. To avoid inconsistency and broken rendering,
+  // we normalize all lists to put the quadpoints in the same standard order
+  // (see https://stackoverflow.com/a/10729881).
```

Comment:
Really good comment!

However, the fact that all of this special-casing is even necessary feels somewhat sad; why is it that Annotations in particular seem to deviate *so much* from the specification in many cases :-(

Reply:
I feel the same. I can only guess that implementation errors were made before and fixing it properly would mean breaking existing PDF files, so they had to go with a backwards compatible solution, forcing other tools to also consider those cases. However, the fact that Adobe applications violate their own specification also gets a sad face from me :-(